test/k8sT/Policies.go

```diff
@@ -808,3 +815,284 @@ var _ = Describe("K8sValidatedPolicyTestAcrossNamespaces", func() {
+			Selector: map[string]string{
+				"pod-name": podName,
+			},
+		},
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	return pod, svc
+}
+
+func cleanupServerPodAndService(k *helpers.Kubectl, pod *v1.Pod, service *v1.Service) {
+	By("Cleaning up the server.")
+	err := k.CoreV1().Pods(pod.Namespace).Delete(pod.Name, nil)
+	Expect(err).To(BeNil(), "Terminating containers are not deleted after timeout")
+
+	By("Cleaning up the server's service.")
+	err = k.CoreV1().Services(service.Namespace).Delete(service.Name, nil)
+	Expect(err).To(BeNil(), "Terminating containers are not deleted after timeout")
+}
+
+func createNetworkClientPod(k *helpers.Kubectl, ns, podName string, targetService *v1.Service, dPort int) *v1.Pod {
+	pod, err := k.CoreV1().Pods(ns).Create(&v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: podName,
+			Labels: map[string]string{
+				"pod-name": podName,
+			},
+		},
+		Spec: v1.PodSpec{
+			RestartPolicy: v1.RestartPolicyNever,
+			Containers: []v1.Container{
+				{
+					Name:  fmt.Sprintf("%s-container", podName),
+					Image: "busybox",
+					Args: []string{
+						"/bin/sh",
+						"-c",
+						fmt.Sprintf("for i in $(seq 1 5); do wget -T 8 %s.%s:%d -O - && exit 0 || sleep 1; done; exit 1",
+							targetService.Name, targetService.Namespace, dPort),
+					},
+				},
+			},
+		},
+	})
+
+	Expect(err).NotTo(HaveOccurred())
+
+	return pod
+}
+
+// testCanConnect creates and tests if a given pod can connect to a given
+// service on a specific destination port.
+func testCanConnect(k *helpers.Kubectl, ns, podName string, service *v1.Service, dPort int) {
+	pod := createNetworkClientPod(k, ns, podName, service, dPort)
+	defer func() {
+		By(fmt.Sprintf("Cleaning up the pod %s", podName))
+		err := k.CoreV1().Pods(ns).Delete(pod.Name, nil)
+		Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Pod %s should have been deleted", pod.Name))
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+	defer cancel()
+	err := k.WaitForPodReady(ctx, ns, podName)
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Pod %s should have been ready", pod.Name))
+
+	ctx, cancel = context.WithTimeout(context.Background(), 2*time.Minute)
+	defer cancel()
+	err = k.WaitForPodFinish(ctx, ns, podName)
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Pod %s should have finish successfully", pod.Name))
+
+	ctx, cancel = context.WithTimeout(context.Background(), 2*time.Minute)
+	defer cancel()
+	success, err := k.WaitForPodSuccess(ctx, ns, podName)
+	Expect(err).NotTo(HaveOccurred(), "Pod did not finish as expected.")
+	Expect(success).To(BeTrue(), "Unable to connect to service %s on port %d. (It should)", service.String(), dPort)
+}
+
+// testCannotConnect creates and tests if a given pod can not connect to a given
+// service on a specific destination port.
+func testCannotConnect(k *helpers.Kubectl, ns, podName string, service *v1.Service, dPort int) {
+	pod := createNetworkClientPod(k, ns, podName, service, dPort)
```

Comment:
Given that `createNetworkClientPod` has `Expect` statements, if any step in `createNetworkClientPod` fails, the test will exit and it will leave cruft behind before the `defer` on the next line. Given that in the current setup on Jenkins, since we continue running tests even if one fails, this will result in all subsequent tests failing most likely since there will be cruft left behind by this test. The tests should gracefully cleanup after themselves in the case of failure, especially if we are not running the test framework on Jenkins with `--failFast`. However, I do see the utility in keeping the environment in the state which caused a test failure for local testing. My main concern is for consistency across the tests; in other words, what state should the framework be left in upon failure?

Reply:
Discussed offline, we're all good now :)